1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+```
+
+User has the ability to explicitly assign a mandatory code:
+
+```powershell
+# some script commands
+...
+    # new script block
+    {
+        $tempFile = New-TemporaryFile
+
+        # Explicitly say to PowerShell Engine how release a resource before the runspace close
+        $tempFile.MakeDisposable([DisposableKind]::AtRunspaceClose, { script to remove the temp file }
+        ...
+    }
+```
+
+## Specification
+
+General statement is to teach `PowerShell Engine` to execute a mandatory (finally) code.
+
+> Notice, .Net Runtime does not ensure finalizer/dispose code execution at application shutdown time.
+
+This covers the main scenarios:
+
+- call `Dispose()` for objects with `IDisposable` to avoid resource leaking
+- call mandatory code for objects without `IDisposable` similar to the one that closes a database connection or deletes a temporary file.
+
+_Our approach should be resource-centric._
+It was expressed by @BrucePay in [related issue](https://github.com/PowerShell/PowerShell/issues/6673), then in another form @lzybkr, then @alx9r.
+
+We need to __register__ resources for which it is necessary to execute a mandatory (finally) code.
+
+This covers all scenarios described in this RFC and all other scenarios.
+It makes the `Engine` smart and brings great UX.
+Old scripts and modules will benefit without any change.
+No changes is needed in tools like [PowerShell EditorSyntax](https://github.com/PowerShell/EditorSyntax).
+
+Try to beat any scenario as part of this resource-centric approach and you will see that the approach covers this scenario even `ForEach-Object -Parallel`) automatically (including dot sourcing).
+(In fact, there can certainly be scenarios where this is impossible in principle, like remoting and serialization, but there may be tradeoffs and workarounds)
+
+### Context / Scope
+
+The `Engine` should take into account Context / Scope.
+
+Since variables can be copied to other scopes, the `Engine` must maintain a _usage counter_ during variable assignment / creation operations.
+This adds a bit of complexity, but it should not affect performance, as in most cases, the `Engine` will only have to check a flag and only in very rare cases do extra work.
+Even an `IDisposable` check needs to be done only once.
+Moreover, we can even make a cache for all known types using our `TypeGen.ps1` utility.
+
+Another feature is that native objects can turn around in `PSObject` and the `Engine` should understand this too.
+
+### Annotating types / objects
+
+There are four options for annotating:
+
+- make a cache for all known types using our `TypeGen.ps1` utility
+- use ETS to enhance some types with this feature with default options
+
+    For example, if we define a TempFile type then users will not have to worry about deleting temporary files created by New-TemporaryFile cmdlet at all
+- embed this in the cmdlets so that they issue an already annotated object into the pipeline
```

Comment:
The concept allows customizations.

>  would thus be a massive breaking change

If you know real example please show. This add a value to the discussion. Otherwise it is a speculation only.

Reply:
Patrick gave some concrete examples in https://github.com/PowerShell/PowerShell/issues/6673 on how the opt-in behaviour is problematic.

You can search any number of scripts for known IDisposable types that would largely fall to pieces if this is opt-in.